src/peergos/shared/hamt/Champ.java

```diff
@@ -0,0 +1,524 @@
+                        bitWidth, maxCollisions, hasher, storage, p.right))
+                .thenCompose(one -> Futures.reduceAll(
+                        Arrays.stream(mappings).collect(Collectors.toList()),
+                        one,
+                        (p, e) -> p.left.put(writer, e.key, hasher.apply(e.key), depth, MaybeMultihash.empty(), e.valueHash,
+                                bitWidth, maxCollisions, hasher, storage, p.right),
+                        (a, b) -> a)
+                );
+    }
+
+    private Champ copyAndSetValue(final int setIndex, final int payloadIndex, final MaybeMultihash val) {
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] dst = new HashPrefixPayload[src.length];
+
+        System.arraycopy(src, 0, dst, 0, src.length);
+        HashPrefixPayload existing = dst[setIndex];
+        KeyElement[] updated = new KeyElement[existing.mappings.length];
+        System.arraycopy(existing.mappings, 0, updated, 0, existing.mappings.length);
+        updated[payloadIndex] = new KeyElement(existing.mappings[payloadIndex].key, val);
+        dst[setIndex] = new HashPrefixPayload(updated, null);
+
+        return new Champ(dataMap, nodeMap, dst);
+    }
+
+    private Champ insertIntoPrefix(final int index, final ByteArrayWrapper key, final MaybeMultihash val) {
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] result = new HashPrefixPayload[src.length];
+
+        System.arraycopy(src, 0, result, 0, src.length);
+        KeyElement[] prefix = new KeyElement[src[index].mappings.length + 1];
+        System.arraycopy(src[index].mappings, 0, prefix, 0, src[index].mappings.length);
+        prefix[prefix.length - 1] = new KeyElement(key, val);
+        // ensure canonical structure
+        Arrays.sort(prefix, Comparator.comparing(m -> m.key));
+        result[index] = new HashPrefixPayload(prefix, null);
+
+        return new Champ(dataMap, nodeMap, result);
+    }
+
+    private Champ addNewPrefix(final int bitpos, final ByteArrayWrapper key, final MaybeMultihash val) {
+        final int insertIndex = index(dataMap, bitpos);
+
+        final HashPrefixPayload[] src = this.contents;
+        final HashPrefixPayload[] result = new HashPrefixPayload[src.length + 1];
+
+        System.arraycopy(src, 0, result, 0, insertIndex);
+        System.arraycopy(src, insertIndex, result, insertIndex + 1, src.length - insertIndex);
+        result[insertIndex] = new HashPrefixPayload(new KeyElement[]{new KeyElement(key, val)}, null);
+
+        BitSet newDataMap = BitSet.valueOf(dataMap.toByteArray());
+        newDataMap.set(bitpos);
+        return new Champ(newDataMap, nodeMap, result);
+    }
+
+    private Champ copyAndMigrateFromInlineToNode(final int bitpos, final Pair<Champ, Multihash> node) {
+
+        final int oldIndex = index(dataMap, bitpos);
+        final int newIndex = this.contents.length - 1 - index(nodeMap, bitpos);
```

Comment:
I'm a  bit lost; is this a length or a position in the array?

Reply:
position in the array, counting from the end

Reply:
(As described in the champ paper)